src/vidpf.rs

```diff
@@ -0,0 +1,667 @@
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    /// y is the sharing of the output.
+    pub y: R,
+    /// pi is a proof used to verify the share.
+    pub pi: Proof,
+}
+
+/// Represents the minimum number of arithmetic operations of the
+/// codomain of a Distributed Point Function (DPF).
+pub trait RangeOps<R>: Sized + Add<Output = R> + Sub<Output = R> {}
+
+/// Represents the codomain of a Distributed Point Function (DPF).
+pub trait Range: PartialEq + ConditionallyNegatable + ConditionallySelectable
+where
+    for<'a> &'a Self: RangeOps<Self>,
+{
+    /// Creates a new value initialized to zero.
+    fn new() -> Self;
+    /// Returns true if the value is zero.
+    fn is_zero(self) -> bool;
+    /// Fills the value from a pseudorandom generator.
+    fn from_seed_stream<R: RngCore>(rng: R) -> Self;
+}
+
+#[derive(Debug)]
+/// `CorrectionWord` adjusts values of shares during VIDPF evaluation.
+pub struct CorrectionWord<R>
+where
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    s_cw: Seed,
+    t_cw_l: ControlBit,
+    t_cw_r: ControlBit,
+    w_cw: R,
+}
+
+#[derive(Debug, PartialEq, Eq, Default)]
+/// Proof
+pub struct Proof([u8; 32]);
+
+impl BitXor for &Proof {
+    type Output = Proof;
+
+    fn bitxor(self, rhs: Self) -> Self::Output {
+        let mut out = Proof::default();
+        zip(&mut out.0, zip(self.0, rhs.0)).for_each(|(c, (a, b))| *c = a ^ b);
+
+        out
+    }
+}
+
+#[derive(Debug)]
+/// `PublicKey` is the common public key of aggregation servers.
+pub struct PublicKey<R>
+where
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    cw: Vec<CorrectionWord<R>>,
+    cs: Vec<Proof>,
+}
+
+#[derive(Debug)]
+/// Key is the private key of an aggregation server.
+pub struct Key {
+    id: ServerID,
+    value: [u8; 16],
+}
+
+impl Key {
+    /// Generates a key at random.
+    ///
+    /// # Errors
+    /// Triggers an error if the random generator fails.
+    pub fn gen(id: ServerID) -> Result<Self, VdafError> {
+        let mut value = [0; 16];
+        getrandom::getrandom(&mut value)?;
+        Ok(Self { id, value })
+    }
+}
+
+struct Sequence(Seed, ControlBit, Seed, ControlBit);
+
+impl BitXor for &Sequence {
+    type Output = Sequence;
+
+    fn bitxor(self, rhs: Self) -> Self::Output {
+        Sequence(
+            &self.0 ^ &rhs.0,
+            self.1 ^ rhs.1,
+            &self.2 ^ &rhs.2,
+            self.3 ^ rhs.3,
+        )
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+/// Stores a boolean value that controls correction operations.
+pub struct ControlBit(Choice);
+
+impl From<u8> for ControlBit {
+    fn from(b: u8) -> Self {
+        Self(Choice::from(b))
+    }
+}
+
+impl From<&Key> for ControlBit {
+    fn from(k: &Key) -> Self {
+        Self::from(k.id as u8)
+    }
+}
+
+impl ConditionallySelectable for ControlBit {
+    fn conditional_select(a: &Self, b: &Self, choice: Choice) -> Self {
+        Self((a.0 & !choice) | (b.0 & choice))
+    }
+}
+
+impl BitAnd<&Seed> for ControlBit {
```

Comment:
This is delicate code... if we get rid of the `vidpf::Seed` wrapper around `XofSeed`, then it might be possible to reuse code from the `idpf` module.

If we cannot reuse code, make sure to add some unit tests for this.

Reply:
changed, it makes things easier.